src/Framework/Framework/Hosting/DotvvmRequestContextExtensions.cs

```diff
@@ -223,24 +230,29 @@ public static string TranslateVirtualPath(string virtualUrl, IHttpContext httpCo
     /// Redirects the client to the specified file.
     /// </summary>
     [Obsolete("Use ReturnFileAsync() instead")]
+    [DoesNotReturn]
     public static void ReturnFile(this IDotvvmRequestContext context, byte[] bytes, string fileName, string mimeType, IEnumerable<KeyValuePair<string, string>>? additionalHeaders = null, string? attachmentDispositionType = null) =>
         context.ReturnFile(new MemoryStream(bytes), fileName, mimeType, additionalHeaders, attachmentDispositionType);
 
     /// <summary>
     /// Redirects the client to the specified file.
     /// </summary>
     [Obsolete("Use ReturnFileAsync() instead")]
+    [DoesNotReturn]
     public static void ReturnFile(this IDotvvmRequestContext context, Stream stream, string fileName, string mimeType, IEnumerable<KeyValuePair<string, string>>? additionalHeaders = null, string? attachmentDispositionType = null) =>
         context.ReturnFileAsync(stream, fileName, mimeType, additionalHeaders, attachmentDispositionType).GetAwaiter().GetResult();
+
     /// <summary>
     /// Redirects the client to the specified file.
     /// </summary>
+    [DoesNotReturn]
```

Comment:
I could not find whether the attribute is valid on async methods, technically it always returns the task instance. However, it shouldn't matter as long as you await it right away